core/src/main/java/org/elasticsearch/cluster/routing/allocation/MoveDecision.java

```diff
@@ -19,109 +19,251 @@
 
 package org.elasticsearch.cluster.routing.allocation;
 
+import org.elasticsearch.cluster.node.DiscoveryNode;
 import org.elasticsearch.cluster.routing.allocation.decider.Decision;
+import org.elasticsearch.cluster.routing.allocation.decider.Decision.Type;
 import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.XContentBuilder;
 
-import java.util.Collections;
-import java.util.Map;
-import java.util.Objects;
+import java.io.IOException;
+import java.util.List;
 
 /**
- * Represents a decision to move a started shard because it is no longer allowed to remain on its current node.
+ * Represents a decision to move a started shard, either because it is no longer allowed to remain on its current node
+ * or because moving it to another node will form a better cluster balance.
  */
-public final class MoveDecision extends RelocationDecision {
+public final class MoveDecision extends AbstractAllocationDecision {
     /** a constant representing no decision taken */
-    public static final MoveDecision NOT_TAKEN = new MoveDecision(null, null, null, null, null);
+    public static final MoveDecision NOT_TAKEN = new MoveDecision(null, null, null, null, null, null, false, 0);
     /** cached decisions so we don't have to recreate objects for common decisions when not in explain mode. */
-    private static final MoveDecision CACHED_STAY_DECISION = new MoveDecision(Decision.YES, Decision.Type.NO, null, null, null);
-    private static final MoveDecision CACHED_CANNOT_MOVE_DECISION = new MoveDecision(Decision.NO, Decision.Type.NO, null, null, null);
+    private static final MoveDecision CACHED_STAY_DECISION =
+        new MoveDecision(null, Decision.YES, null, Type.NO, null, null, false, 0);
+    private static final MoveDecision CACHED_CANNOT_MOVE_DECISION =
+        new MoveDecision(null, Decision.NO, null, Type.NO, null, null, false, 0);
 
     @Nullable
+    private final DiscoveryNode currentNode;
+    @Nullable
     private final Decision canRemainDecision;
     @Nullable
-    private final Map<String, NodeAllocationResult> nodeDecisions;
+    private final Decision canRebalanceDecision;
+    private final boolean fetchPending;
```

Comment:
here we use a boolean fetchPending whereas AllocateUnassignedDecision uses the allocationStatus to determine if fetch is pending. As said in an earlier comment, I think we should have a uniform result object for this.